File: src/lib/mode.ts
import type { ThemeMode } from "@/lib/types";

export const AUTO_APPLY_MODE_CONFIDENCE = 0.9;

type ConfidenceLevel = "high" | "medium" | "low";

type TokenWeight = {
  token: string;
  weight: number;
};

export type ModeInference = {
  mode: ThemeMode;
  confidence: number;
  confidenceLevel: ConfidenceLevel;
  reason: string;
};

export type ModeHintResolution = {
  mode: ThemeMode;
  matchedKey: string;
  normalizedMatch: boolean;
};

const RE_BASE16_LIGHT = /^base16-.+-light$/i;
const RE_BASE16_DARK = /^base16-(?!.*-light$).+$/i;

const RE_TRAILING_STYLE_MODIFIER =
  /[-_](bold|italic|dim|dimmed|soft|hard|mono|minimal|default|main|opaque|transparent|contrast|highcontrast)$/;

const LIGHT_TOKENS: readonly TokenWeight[] = [
  { token: "light", weight: 3 },
  { token: "day", weight: 3 },
  { token: "dawn", weight: 3 },
  { token: "operandi", weight: 3 },
  { token: "latte", weight: 3 },
  { token: "sun", weight: 3 },
  { token: "morning", weight: 2 },
  { token: "white", weight: 2 },
  { token: "paper", weight: 2 },
  { token: "cream", weight: 2 },
  { token: "snow", weight: 2 },
  { token: "lumi", weight: 1 },
  { token: "bright", weight: 1 },
] as const;

const DARK_TOKENS: readonly TokenWeight[] = [
  { token: "dark", weight: 3 },
  { token: "night", weight: 3 },
  { token: "vivendi", weight: 3 },
  { token: "moon", weight: 3 },
  { token: "storm", weight: 3 },
  { token: "mocha", weight: 3 },
  { token: "frappe", weight: 3 },
  { token: "macchiato", weight: 3 },
  { token: "shadow", weight: 2 },
  { token: "midnight", weight: 2 },
  { token: "abyss", weight: 2 },
  { token: "dusk", weight: 2 },
  { token: "deep", weight: 1 },
  { token: "black", weight: 1 },
  { token: "dragon", weight: 1 },
  { token: "wave", weight: 1 },
  { token: "ember", weight: 1 },
  { token: "fog", weight: 1 },
  { token: "moss", weight: 1 },
] as const;

const COMPACT_LIGHT_EXCLUSIONS = new Set([
  "twilight",
  "starlight",
  "moonlight",
  "spotlight",
  "limelight",
  "highlight",
] as const);

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function normalizeName(value: string): string {
  return value.trim().toLowerCase();
}

function stripTrailingStyleModifiers(value: string): string {
  let output = value;
  while (RE_TRAILING_STYLE_MODIFIER.test(output)) {
    output = output.replace(RE_TRAILING_STYLE_MODIFIER, "");
  }
  return output;
}

function hasBoundaryToken(text: string, token: string): boolean {
  return new RegExp(`(^|[-_])${escapeRegex(token)}($|[-_])`, "i").test(text);
}

function hasTerminalBoundaryToken(text: string, token: string): boolean {
  return new RegExp(`(^|[-_])${escapeRegex(token)}$`, "i").test(text);
}

function hasCompactTerminalLight(text: string): boolean {
  if (!text.endsWith("light")) return false;
  if (text.endsWith("-light") || text.endsWith("_light")) return false;
  for (const excluded of COMPACT_LIGHT_EXCLUSIONS) {
    if (text.endsWith(excluded)) return false;
  }
  return true;
}

function hasCompactTerminalDark(text: string): boolean {
  if (!text.endsWith("dark")) return false;
  if (text.endsWith("-dark") || text.endsWith("_dark")) return false;
  return true;
}

function scoreTokens(
  text: string,
  tokens: readonly TokenWeight[],
): { score: number; strongestTerminal: number; labels: string[] } {
  let score = 0;
  let strongestTerminal = 0;
  const labels: string[] = [];

  for (const { token, weight } of tokens) {
    if (!hasBoundaryToken(text, token)) continue;
    score += weight;
    labels.push(token);
    if (hasTerminalBoundaryToken(text, token)) {
      strongestTerminal = Math.max(strongestTerminal, weight);
    }
  }

  return { score, strongestTerminal, labels };
}

function confidenceFromScores(
  dominant: number,
  runnerUp: number,
): { value: number; level: ConfidenceLevel } {
  const delta = dominant - runnerUp;
  if (dominant >= 3 && delta >= 2) {
    return { value: 0.95, level: "high" };
  }
  if (dominant >= 2 && delta >= 1) {
    return { value: 0.8, level: "medium" };
  }
  return { value: 0.65, level: "low" };
}

export function inferThemeMode(name: string | undefined): ModeInference | undefined {
  if (!name) return undefined;

  const normalized = normalizeName(name);
  if (!normalized) return undefined;

  if (RE_BASE16_LIGHT.test(normalized)) {
    return {
      mode: "light",
      confidence: 1,
      confidenceLevel: "high",
      reason: "Base16 light variant pattern",
    };
  }

  if (RE_BASE16_DARK.test(normalized)) {
    return {
      mode: "dark",
      confidence: 0.95,
      confidenceLevel: "high",
      reason: "Base16 dark variant pattern",
    };
  }

  const stripped = stripTrailingStyleModifiers(normalized);
  const light = scoreTokens(stripped, LIGHT_TOKENS);
  const dark = scoreTokens(stripped, DARK_TOKENS);

  if (hasCompactTerminalLight(stripped)) {
    light.score += 2;
    light.strongestTerminal = Math.max(light.strongestTerminal, 2);
    light.labels.push("light(compact)");
  }

  if (hasCompactTerminalDark(stripped)) {
    dark.score += 2;
    dark.strongestTerminal = Math.max(dark.strongestTerminal, 2);
    dark.labels.push("dark(compact)");
  }

  if (light.score === 0 && dark.score === 0) {
    return undefined;
  }

  if (light.score > 0 && dark.score > 0) {
    const terminalDelta = light.strongestTerminal - dark.strongestTerminal;
    if (terminalDelta >= 2) {
      const confidence = confidenceFromScores(light.score, dark.score);
      return {
        mode: "light",
        confidence: confidence.value,
        confidenceLevel: confidence.level,
        reason: `Light tokens dominate (${light.labels.join(", ")})`,
      };
    }
    if (terminalDelta <= -2) {
      const confidence = confidenceFromScores(dark.score, light.score);
      return {
        mode: "dark",
        confidence: confidence.value,
        confidenceLevel: confidence.level,
        reason: `Dark tokens dominate (${dark.labels.join(", ")})`,
      };
    }
    return undefined;
  }

  const mode: ThemeMode = light.score > dark.score ? "light" : "dark";
  const dominant = mode === "light" ? light : dark;
  const runnerUp = mode === "light" ? dark.score : light.score;
  const confidence = confidenceFromScores(dominant.score, runnerUp);

  return {
    mode,
    confidence: confidence.value,
    confidenceLevel: confidence.level,
    reason: `${mode === "light" ? "Light" : "Dark"} tokens matched (${dominant.labels.join(", ")})`,
  };
}

export function normalizeModeHintKey(name: string): string {
  return normalizeName(name).replace(/[^a-z0-9]/g, "");
}

export function mergeModeHintRecords(
  repo: string,
  existing: Record<string, ThemeMode>,
  incoming: Record<string, ThemeMode>,
): Record<string, ThemeMode> {
  const merged = { ...existing };
  const index = new Map<string, string>();

  for (const key of Object.keys(merged)) {
    index.set(normalizeModeHintKey(key), key);
  }

  for (const [key, mode] of Object.entries(incoming)) {
    const normalized = normalizeModeHintKey(key);
    const existingKey = index.get(normalized);
    if (!existingKey) {
      merged[key] = mode;
      index.set(normalized, key);
      continue;
    }

    if (merged[existingKey] !== mode) {
      throw new Error(
        `Conflicting mode hints for ${repo}/${key}: ${merged[existingKey]} vs ${mode}`,
      );
    }
  }

  return merged;
}

export function resolveModeHint(
  variantName: string,
  hints: Record<string, ThemeMode>,
): ModeHintResolution | undefined {
  const exact = hints[variantName];
  if (exact) {
    return { mode: exact, matchedKey: variantName, normalizedMatch: false };
  }

  const normalizedName = normalizeModeHintKey(variantName);
  for (const [hintName, mode] of Object.entries(hints)) {
    if (normalizeModeHintKey(hintName) === normalizedName) {
      return { mode, matchedKey: hintName, normalizedMatch: true };
    }
  }

  return undefined;
}

File: internal/theme/mode.go
package theme

import (
	"errors"
	"fmt"
	"sort"
	"strings"
)

// Ported verbatim from src/lib/mode.ts. Heuristics must not drift: the
// golden fixtures under testdata/golden/expected/functions are the oracle.

const AutoApplyModeConfidence = 0.9

// ModeInference is the outcome of inferring a brightness mode from a name.
type ModeInference struct {
	Mode       Mode    `json:"mode"`
	Confidence float64 `json:"confidence"`
	Level      string  `json:"confidenceLevel"` // high | medium | low
	Reason     string  `json:"reason"`
}

// ResolvedHint is a mode-hint lookup result for a variant name.
type ResolvedHint struct {
	Mode            Mode   `json:"mode"`
	MatchedKey      string `json:"matchedKey"`
	NormalizedMatch bool   `json:"normalizedMatch"`
}

type tokenWeight struct {
	token  string
	weight int
}

var lightTokens = []tokenWeight{
	{"light", 3}, {"day", 3}, {"dawn", 3}, {"operandi", 3}, {"latte", 3},
	{"sun", 3}, {"morning", 2}, {"white", 2}, {"paper", 2}, {"cream", 2},
	{"snow", 2}, {"lumi", 1}, {"bright", 1},
}

var darkTokens = []tokenWeight{
	{"dark", 3}, {"night", 3}, {"vivendi", 3}, {"moon", 3}, {"storm", 3},
	{"mocha", 3}, {"frappe", 3}, {"macchiato", 3}, {"shadow", 2},
	{"midnight", 2}, {"abyss", 2}, {"dusk", 2}, {"deep", 1}, {"black", 1},
	{"dragon", 1}, {"wave", 1}, {"ember", 1}, {"fog", 1}, {"moss", 1},
}

var compactLightExclusions = map[string]bool{
	"twilight": true, "starlight": true, "moonlight": true,
	"spotlight": true, "limelight": true, "highlight": true,
}

// stripTrailingStyleModifiers replicates
// /[-_](bold|italic|dim|dimmed|soft|hard|mono|minimal|default|main|opaque|
// transparent|contrast|highcontrast)$/ applied repeatedly.
var styleModifiers = []string{
	"bold", "italic", "dim", "dimmed", "soft", "hard", "mono", "minimal",
	"default", "main", "opaque", "transparent", "contrast", "highcontrast",
}

func stripTrailingStyleModifiers(value string) string {
	for {
		stripped := false
		for _, mod := range styleModifiers {
			suffix := "_" + mod
			if strings.HasSuffix(value, suffix) || strings.HasSuffix(value, "-"+mod) {
				if len(value) > len(suffix) {
					value = value[:len(value)-len(suffix)]
					stripped = true
					break
				}
			}
		}
		if !stripped {
			return value
		}
	}
}

// hasBoundaryToken replicates new RegExp(`(^|[-_])token($|[-_])`).test(text).
func hasBoundaryToken(text, token string) bool {
	for i := 0; ; {
		idx := strings.Index(text[i:], token)
		if idx < 0 {
			return false
		}
		start := i + idx
		end := start + len(token)
		startOK := start == 0 || text[start-1] == '-' || text[start-1] == '_'
		endOK := end == len(text) || text[end] == '-' || text[end] == '_'
		if startOK && endOK {
			return true
		}
		i = start + 1
	}
}

func hasTerminalBoundaryToken(text, token string) bool {
	if !strings.HasSuffix(text, token) {
		return false
	}
	head := text[:len(text)-len(token)]
	return head == "" || head[len(head)-1] == '-' || head[len(head)-1] == '_'
}

func hasCompactTerminalLight(text string) bool {
	if !strings.HasSuffix(text, "light") {
		return false
	}
	if strings.HasSuffix(text, "-light") || strings.HasSuffix(text, "_light") {
		return false
	}
	for excluded := range compactLightExclusions {
		if strings.HasSuffix(text, excluded) {
			return false
		}
	}
	return true
}

func hasCompactTerminalDark(text string) bool {
	if !strings.HasSuffix(text, "dark") {
		return false
	}
	return !strings.HasSuffix(text, "-dark") && !strings.HasSuffix(text, "_dark")
}

type tokenScore struct {
	score             int
	strongestTerminal int
	labels            []string
}

func scoreTokens(text string, tokens []tokenWeight) tokenScore {
	ts := tokenScore{labels: []string{}}
	for _, tw := range tokens {
		if !hasBoundaryToken(text, tw.token) {
			continue
		}
		ts.score += tw.weight
		ts.labels = append(ts.labels, tw.token)
		if hasTerminalBoundaryToken(text, tw.token) && tw.weight > ts.strongestTerminal {
			ts.strongestTerminal = tw.weight
		}
	}
	return ts
}

func confidenceFromScores(dominant, runnerUp int) (float64, string) {
	delta := dominant - runnerUp
	switch {
	case dominant >= 3 && delta >= 2:
		return 0.95, "high"
	case dominant >= 2 && delta >= 1:
		return 0.8, "medium"
	default:
		return 0.65, "low"
	}
}

// InferThemeMode ports src/lib/mode.ts inferThemeMode. Returns nil when the
// TS implementation returned undefined (empty name or no signal at all).
func InferThemeMode(name string) *ModeInference {
	normalized := strings.ToLower(strings.TrimSpace(name))
	if normalized == "" {
		return nil
	}

	const b16 = "base16-"
	isBase16 := strings.HasPrefix(normalized, b16)
	endsLight := strings.HasSuffix(normalized, "-light")
	// /^base16-.+-light$/ : at least one char between prefix and -light.
	if isBase16 && endsLight && len(normalized) > len(b16)+len("-light") {
		return &ModeInference{Mode: ModeLight, Confidence: 1, Level: "high", Reason: "Base16 light variant pattern"}
	}
	// /^base16-(?!.*-light$).+$/ : non-empty remainder that does not end -light.
	if isBase16 && !endsLight && len(normalized) > len(b16) {
		return &ModeInference{Mode: ModeDark, Confidence: 0.95, Level: "high", Reason: "Base16 dark variant pattern"}
	}

	stripped := stripTrailingStyleModifiers(normalized)
	light := scoreTokens(stripped, lightTokens)
	dark := scoreTokens(stripped, darkTokens)

	if hasCompactTerminalLight(stripped) {
		light.score += 2
		if light.strongestTerminal < 2 {
			light.strongestTerminal = 2
		}
		light.labels = append(light.labels, "light(compact)")
	}
	if hasCompactTerminalDark(stripped) {
		dark.score += 2
		if dark.strongestTerminal < 2 {
			dark.strongestTerminal = 2
		}
		dark.labels = append(dark.labels, "dark(compact)")
	}

	if light.score == 0 && dark.score == 0 {
		return nil
	}

	if light.score > 0 && dark.score > 0 {
		terminalDelta := light.strongestTerminal - dark.strongestTerminal
		if terminalDelta >= 2 {
			conf, level := confidenceFromScores(light.score, dark.score)
			return &ModeInference{
				Mode: ModeLight, Confidence: conf, Level: level,
				Reason: fmt.Sprintf("Light tokens dominate (%s)", strings.Join(light.labels, ", ")),
			}
		}
		if terminalDelta <= -2 {
			conf, level := confidenceFromScores(dark.score, light.score)
			return &ModeInference{
				Mode: ModeDark, Confidence: conf, Level: level,
				Reason: fmt.Sprintf("Dark tokens dominate (%s)", strings.Join(dark.labels, ", ")),
			}
		}
		return nil
	}

	mode := ModeDark
	if light.score > dark.score {
		mode = ModeLight
	}
	dominant, runnerUp := dark, light
	if mode == ModeLight {
		dominant, runnerUp = light, dark
	}
	conf, level := confidenceFromScores(dominant.score, runnerUp.score)
	label := "Dark"
	if mode == ModeLight {
		label = "Light"
	}
	return &ModeInference{
		Mode: mode, Confidence: conf, Level: level,
		Reason: fmt.Sprintf("%s tokens matched (%s)", label, strings.Join(dominant.labels, ", ")),
	}
}

// NormalizeModeHintKey lowercases and keeps only [a-z0-9].
func NormalizeModeHintKey(name string) string {
	var b strings.Builder
	for _, r := range strings.ToLower(strings.TrimSpace(name)) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// MergeModeHintRecords folds incoming hints into existing, keyed by their
// normalized form. Conflicting modes for the same normalized key are an error.
func MergeModeHintRecords(repo string, existing, incoming map[string]Mode) (map[string]Mode, error) {
	merged := make(map[string]Mode, len(existing)+len(incoming))
	index := make(map[string]string, len(existing)+len(incoming))
	for k, v := range existing {
		merged[k] = v
		index[NormalizeModeHintKey(k)] = k
	}

	keys := make([]string, 0, len(incoming))
	for k := range incoming {
		keys = append(keys, k)
	}
	sort.Strings(keys) // deterministic conflict order

	for _, k := range keys {
		mode := incoming[k]
		normalized := NormalizeModeHintKey(k)
		existingKey, hit := index[normalized]
		if !hit {
			merged[k] = mode
			index[normalized] = k
			continue
		}
		if merged[existingKey] != mode {
			return nil, fmt.Errorf("Conflicting mode hints for %s/%s: %s vs %s", repo, k, merged[existingKey], mode)
		}
	}
	return merged, nil
}

// ResolveModeHint looks up an exact hint first, then a normalized match.
func ResolveModeHint(variantName string, hints map[string]Mode) *ResolvedHint {
	if mode, ok := hints[variantName]; ok {
		return &ResolvedHint{Mode: mode, MatchedKey: variantName, NormalizedMatch: false}
	}
	normalized := NormalizeModeHintKey(variantName)
	if normalized == "" {
		return nil
	}
	keys := make([]string, 0, len(hints))
	for k := range hints {
		keys = append(keys, k)
	}
	sort.Strings(keys) // deterministic match order
	for _, k := range keys {
		if NormalizeModeHintKey(k) == normalized {
			return &ResolvedHint{Mode: hints[k], MatchedKey: k, NormalizedMatch: true}
		}
	}
	return nil
}

var errInvalidRepoPayload = errors.New("invalid repository payload")

// InferModeFromColorscheme returns nil below the auto-apply threshold (0.9).
func InferModeFromColorscheme(colorscheme string) *Mode {
	inference := InferThemeMode(colorscheme)
	if inference == nil || inference.Confidence < AutoApplyModeConfidence {
		return nil
	}
	mode := inference.Mode
	return &mode
}
